[QOL] Rehash Angel Block tweak using "floatato technology"

axialeaa opened the pull request:
Hi! I'd like to propose an alternative implementation of the angel block tweak inspired by the mechanics of the most recent April Fool's snapshot's "floatato".
***

### Description
In the current push of Tweakeroo, angel blocks are set to place one block ahead of the player in their horizontal facing direction (only taking yaw into account). Since players (unlike blocks) are unbounded by the world grid coordinates, this can often lead to the block being placed in an uncomfortable/unintuitive spot relative to the camera, often obstructing the view.

Mojang's implementation of the floatato gets around this issue by drawing a vector from the player's eye position to exactly 3.0 units ahead in their exact facing direction (taking into account both pitch and yaw) and placing the floatato there. 3.0 is just a bit less than the maximum reach distance, allowing players to comfortably bridge in the sky without sacrificing practicality.

This pull request contains similar logic, albeit fitting in-line with tweakeroo's existing feature set. For that reason, I've added a `ConfigDouble` slider to customise the vector length. By default it's set to 3.0, but it can go up to 5.0 which seems to be the maximum before the server starts rejecting packets, even if the blockReachDistance option is set higher. In theory this would be nice to find a workaround for, but it's not the end of the world.
***

### Other Related Changes & Fixes
- Fixed the player's hand failing to swing when placing angel blocks
- (Potentially) fixed an anecdotal lag spike when placing angel blocks
- Added nullchecks for `mc.world` and `mc.interactionManager`
- Added `IMixinEntity` accessor to invoke `Entity#getRotationVec(float, float)`

axialeaa committed:
added angelBlockPlacementDistance
reimplemented angel block using floatato technology

axialeaa committed:
changed min angelBlockPlacementDistance value to 1. pretty much anything less than that intersects with the player's collision box and therefore causes the placement attempt to fail

maruohon commented on src/main/java/fi/dy/masa/tweakeroo/config/Configs.java:
```diff
@@ -1,23 +1,10 @@
-import fi.dy.masa.malilib.config.options.ConfigString;
-import fi.dy.masa.malilib.config.options.ConfigStringList;
+import fi.dy.masa.malilib.config.*;
+import fi.dy.masa.malilib.config.options.*;
```
Can you revert all the import style and ordering changes? This doesn't match my settings and personal preferences.

maruohon commented on src/main/java/fi/dy/masa/tweakeroo/event/InputHandler.java:
```diff
@@ -94,36 +88,35 @@ public boolean onMouseClick(int mouseX, int mouseY, int eventButton, boolean eve
-                Vec3d hitVec = PositionUtils.getHitVecCenter(posFront, facing);
-                BlockHitResult context = new BlockHitResult(hitVec, facing, posFront, false);
-                ItemStack stack = mc.player.getMainHandStack();
+        if (GuiUtils.getCurrentScreen() != null)
```
Also combine this with the previous if

maruohon commented on src/main/java/fi/dy/masa/tweakeroo/event/InputHandler.java:
```diff
@@ -94,36 +88,35 @@ public boolean onMouseClick(int mouseX, int mouseY, int eventButton, boolean eve
+            for (Hand hand : Hand.values())
+            {
+                ItemStack stack = mc.player.getStackInHand(hand);
+                if (!stack.isEmpty() && stack.getItem() instanceof BlockItem)
```
While pretty much everyone else hates it, my style is to use `foo == false` instead of `!foo` in all of my MC mods.
Also I use an empty like before and after any if/switch/for/etc. unless it's the first/last thing inside the method.

maruohon commented on src/main/java/fi/dy/masa/tweakeroo/event/InputHandler.java:
```diff
@@ -94,36 +88,35 @@ public boolean onMouseClick(int mouseX, int mouseY, int eventButton, boolean eve
+        if (mc.player.isCreative() && FeatureToggle.TWEAK_ANGEL_BLOCK.getBooleanValue() && eventButtonState && mc.options.useKey.matchesMouse(eventButton) && mc.crosshairTarget.getType() == HitResult.Type.MISS)
+        {
+            Vec3d eyePos = mc.player.getEyePos();
+            Vec3d rotVec = ((IMixinEntity) mc.player).tweakeroo_getRotationVector(mc.player.getPitch(), mc.player.getYaw());
```
There is no need to add this `@Invoker`, you can instead use the following method from `Entity`:
```java
public final Vec3d getRotationVec(float tickDelta)
```
So basically just call `getRotationVec(1.0f);`.

maruohon commented on src/main/resources/mixins.tweakeroo.json:
```diff
@@ -1,82 +1,83 @@
-	"injectors": {
-		"defaultRequire": 0
-	}
+  "required": true,
```
Since the `@Invoker` is not needed, you can revert this file entirely. In case it was needed, you should avoid making these whitespace changes.

axialeaa committed:
removed accessor mixin and inhabiting @Invoker (now uses `Entity#getRotationVec(float)`)
reverted accidental import shuffling (thanks intellij idea)
whitespace/formatting changes
!foo -> foo == false
combined if statements with identical returns

axialeaa committed:
rolled back mixins.tweakeroo.json

axialeaa commented:
Those things should now be fixed. Thank you for responding.
The import ordering and mixin whitespace changes were not because of me, but Intellij IDEA having a mind of its own lol

maruohon commented:
Thanks. Yeah the import ordering depends on your IDEA configs. I have customized the import ordering to my preferences, which is basically other common classes first, then vanilla classes, and last my own mod classes.
